prusti-interface/src/environment/mod.rs

```diff
@@ -303,12 +305,56 @@ impl<'tcx> Environment<'tcx> {
                     base_body: Rc::new(body),
                     monomorphised_bodies: HashMap::new(),
                     borrowck_facts: Rc::new(facts),
+                    monomorphised_bodies_with_caller: HashMap::new(),
                 }
             });
         body
             .monomorphised_bodies
             .entry(substs)
-            .or_insert_with(|| ty::EarlyBinder(body.base_body.clone()).subst(self.tcx, substs))
+            .or_insert_with(|| {
+                let param_env = self.tcx.param_env(def_id);
+                let substituted = ty::EarlyBinder(body.base_body.clone()).subst(self.tcx, substs);
+                self.resolve_assoc_types(substituted.clone(), param_env)
+            })
+            .clone()
+    }
+
+    pub fn local_mir_with_caller(
+        &self,
+        def_id: LocalDefId,
+        caller_def_id: LocalDefId,
+        substs: SubstsRef<'tcx>,
+    ) -> Rc<mir::Body<'tcx>> {
```

Comment:
@JonasAlaif Will this change be easy to incorporate into your refactored `Environment`? For proper monomorphisation (and assoc. type normalisation) I need called `DefId`, type substitutions (as before), but also the *caller* `DefId`.

And should I wait for your PR to land or can I merge this first (assuming the tests pass)?